Mark Variant type on empty strings invalid as well

bastelfreak opened the pull request.

bastelfreak committed:
Mark Variant type on empty strings invalid as well

bastelfreak committed:
enhance README.md

ekohl commented on README.md:
```diff
@@ -39,22 +39,30 @@ recommendations are documented at [voxpupuli.org](https://voxpupuli.org/docs/rev
 class foo (
   String $bar = '',
+  String[0] $foo = '',
+  Variant $baz = '',
+  Variant[] $foobar = '',
```
Is this even valid? Does this even compile?

bastelfreak replied:
ah it's not. I remove it:
```
 $ cat test.pp
class foo (
  String $bar = '',
  String[0] $foo = '',
  Variant $baz = '',
  Variant[] $foobar = '',
) {
  # logic
}
 $ puppet parser validate test.pp
Error: Could not parse for environment production: Syntax error at ']' (file: /home/bastelfreak/code/puppet-lint_modulesync_configs/modules/voxpupuli/puppet-lint-params_empty_string-check/test.pp, line: 5, column: 11)
 $
```

ekohl reviewed (commented):
In puppetlabs-postgresql I saw a nice example. `String = 'public'`. It needed to handle empty too (to indicate there is no schema) so effectively it must be `String[0] = 'public'`. Is that also something the plugin could handle or do you need control statements?

Speaking of control statements, I don't see an example of that in `Disabling the check`.

bastelfreak commented:
@ekohl I added a check for `String[0] = 'public'`. This is allowed. The purpose of this plugin is to focus in parameters with `''` as default value. I also removed the bogus `Variant[] $foobar = '',` example as that's already a parser error.

ekohl commented on README.md:
```diff
@@ -39,22 +39,29 @@ recommendations are documented at [voxpupuli.org](https://voxpupuli.org/docs/rev
 class foo (
   String $bar = '',
+  String[0] $foo = '',
+  Variant $baz = '',
```
I'm still surprised this is legal. I've never seen code that actually has this. `Variant $baz` doesn't actually say anything. If it's accepted by Puppet, I think that deserves a bug report.

bastelfreak replied:
yes. I raised https://tickets.puppetlabs.com/browse/PUP-11366

ekohl commented on README.md:
````diff
@@ -39,22 +39,29 @@ recommendations are documented at [voxpupuli.org](https://voxpupuli.org/docs/rev
 }
 ```
 
+All four of the above variantions aren't recommended.
````
It made me think of https://www.youtube.com/watch?v=jk3EsXgXcyQ, but in this case there are 3 ;)

bastelfreak replied:
fixed it

ekohl commented on lib/puppet-lint/plugins/check-params_empty_string_assignment.rb:
```diff
@@ -12,7 +12,7 @@ def check
         next if default_value.empty?
         # if the parameter has datatype String/String[] and a value of ''
-        if type[0].type == :TYPE && type[0].value =~ /^String/ && default_value[0].value == ''
+        if type[0].type == :TYPE && type[0].value =~ /^(String|Variant)/ && default_value[0].value == ''
```
I feel like at this point it would be useful to have even richer parsing so you don't manually need to parse `String[0]` into `String` and `0`. Perhaps worth an issue to `puppet-lint` itself?

bastelfreak replied:
that would help if we differentiate between `String` and `String[0]`, but that's currently not the case. Instead of the regex I could do an explicit match against all types, but I think this is sufficient (because we don't need to differentiate between the non-allowed types)

bastelfreak commented:
I updated the README.md with a control comment. I also tested it locally to ensure it actually works.

alexjfisher commented on README.md:
````diff
@@ -39,22 +39,28 @@ recommendations are documented at [voxpupuli.org](https://voxpupuli.org/docs/rev
 }
 ```
 
+All three of the above variantions aren't recommended.
````
```suggestion
All three of the above variations aren't recommended
```
